src/components/CreditCardNumberInput/CreditCardNumberInput.tsx

```diff
@@ -0,0 +1,74 @@
+import React, { useState, useEffect, useCallback, useMemo } from 'react';
+import InputMask, { Props as InputMaskProps } from 'react-input-mask';
+import creditCardType from 'credit-card-type';
+import { Field, FieldProps } from '../Field';
+
+export type Props = Omit<InputMaskProps & FieldProps, 'mask' | 'value' | 'defaultValue'> & {
+  value?: string;
+  defaultValue?: string;
+};
+
+const DEFAULT_MASK = '9999 9999 9999 9999';
+
+export const CreditCardNumberInput: React.FC<Props> = ({
+  value: propsValue,
+  disabled,
+  readOnly,
+  defaultValue,
+  maskChar,
+  alwaysShowMask,
+  onMouseDown,
+  onFocus,
+  onBlur,
+  onChange,
+  ...rest
+}) => {
+  const [value, setValue] = useState(propsValue ?? defaultValue ?? '');
+
+  const mask = useMemo(() => {
+    const cardType = creditCardType(value)[0];
+    if (!cardType) return DEFAULT_MASK;
+
+    const { gaps, lengths } = cardType;
+    if (!gaps || !lengths) return DEFAULT_MASK;
+
+    const length = lengths[lengths.length - 1];
+    const digits = Array(length).fill('9');
+    return [...gaps]
+      .reverse()
+      .reduce(
+        (mask, gapIndex) => [...mask.slice(0, gapIndex), ' ', ...mask.slice(gapIndex)],
```

Comment:
I wonder if a reduceRight here with some fancy indexing math would also work in order to get rid of the reverse

Reply:
Yeah actually `reduceRight` should work here, let me try.